packages/core/src/components/overlay/overlay.tsx

```diff
@@ -348,7 +349,19 @@ export class Overlay extends React.PureComponent<IOverlayProps, IOverlayState> {
     private handleDocumentClick = (e: MouseEvent) => {
         const { isOpen, onClose } = this.props;
         const eventTarget = e.target as HTMLElement;
-        const isClickInOverlay = this.containerElement != null && this.containerElement.contains(eventTarget);
+
+        const { openStack } = Overlay;
+        const stackIndex = openStack.indexOf(this);
+
+        const isClickInDescendantOverlay = openStack
+            .slice(stackIndex)
+            .map(findDOMNode)
+            .filter(elem => elem && elem.contains)
+            .some(elem => elem.contains(eventTarget));
+
+        const isClickInOverlay =
+            (this.containerElement != null && this.containerElement.contains(eventTarget)) ||
+            isClickInDescendantOverlay;
```

Comment:
don't `||` here cuz it makes the variable name incorrect (includes a check for clicks _not_ in overlay). `||` in the `if`.